.github/workflows/ci.yml

```diff
@@ -49,13 +50,147 @@ jobs:
+          else
+            variants+=("-cnb" "-cnb-build")
+          fi
+          for variant in "${variants[@]}"; do
+            srcTag="heroku/heroku:${{ matrix.stack-version}}${variant}"
+            destTag="${srcTag}${platformSuffix}${TAG_SUFFIX}"
+            docker tag "${srcTag}" "${destTag}"
+            docker push "${destTag}"
+          done
+
+  publish-indices:
+    if: github.ref_name == 'main' || github.ref_type == 'tag'
+    name: "Publish heroku-${{ matrix.stack-version }} indices"
+    needs:
+      - publish-images
+    runs-on: ubuntu-24.04
+    env:
+      TAG_SUFFIX: ".${{ github.ref_type == 'tag' && github.ref_name || 'nightly' }}"
+    strategy:
+      fail-fast: false
+      matrix:
+        stack-version: ["24"]
+    steps:
+      - name: Log into Docker Hub
+        run: echo '${{ secrets.DOCKER_HUB_TOKEN }}' | docker login -u '${{ secrets.DOCKER_HUB_USERNAME }}' --password-stdin
+      - name: Publish multi-arch image index
+        run: |
+          for variant in '' '-build'; do
+            indexTag="heroku/heroku:${{ matrix.stack-version }}${variant}${TAG_SUFFIX}"
+            armTag="heroku/heroku:${{ matrix.stack-version }}${variant}_linux-arm64${TAG_SUFFIX}"
+            amdTag="heroku/heroku:${{ matrix.stack-version }}${variant}_linux-amd64${TAG_SUFFIX}"
+            docker manifest create "$indexTag" "$amdTag" "$armTag"
+            docker manifest push "$indexTag"
+          done
+
+  promote-tags:
+    if: github.ref_type == 'tag'
+    name: "Promote heroku-${{ matrix.stack-version }} tags"
+    needs:
+      - publish-images
+      - publish-indices
+    runs-on: ubuntu-24.04
+    strategy:
+      fail-fast: false
+      matrix:
+        stack-version: ["20", "22", "24"]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Log into Docker Hub
+        run: echo '${{ secrets.DOCKER_HUB_TOKEN }}' | docker login -u '${{ secrets.DOCKER_HUB_USERNAME }}' --password-stdin
+      - name: Install crane
+        uses: buildpacks/github-actions/setup-tools@v5.7.3
+      - name: Promote images to stable tag
+        run: |
+          destTags=( )
+          if (( ${{ matrix.stack-version }} >= 24 )); then
+            for variant in '' '-build'; do
+              for arch in 'amd64' 'arm64'; do
+                destTags+=("heroku/heroku:${{ matrix.stack-version }}${variant}_linux-${arch}")
+              done
+              destTags+=("heroku/heroku:${{ matrix.stack-version }}${variant}")
+            done
+          else
+            for variant in '' '-build' '-cnb' '-cnb-build'; do
+              destTags+=("heroku/heroku:${{ matrix.stack-version }}${variant}")
+            done
+          fi
+          for destTag in "${destTags[@]}"; do
+            srcTag="${destTag}.${{ github.ref_name }}"
+            crane copy "${srcTag}" "${destTag}"
+          done
+
+  publish-stack:
+    if: github.ref_type == 'tag'
+    name: "Publish heroku-${{ matrix.stack-version }} stack"
```

Comment:
Updated in a few places to read more like "Upload heroku-* runtime images". I tried to use parlance like "heroku runtime images" to clarify the difference between the OCI images and the cheverny type images. It's a challenge because "stack" and "image" describe both concepts. Also "manifest" is both a term for OCI images, and a different thing we upload to cheverny.

Reply:
I like "runtime images" - IMO that works well enough. And agree hard to juggle all of these overloaded terms :-)